packages/test/local-server-tests/src/test/synchronousDataStoreCreation.spec.ts

```diff
@@ -0,0 +1,207 @@
+	}
+}
+
+// a simple datastore factory that is also a registry so that it can create instances of itself
+class DataStoreWithSyncCreateFactory
+	implements IFluidDataStoreFactory, IFluidDataStoreRegistry
+{
+	static readonly instance = new DataStoreWithSyncCreateFactory();
+	public readonly type = DataStoreWithSyncCreate.type;
+
+	private constructor() {}
+
+	get IFluidDataStoreRegistry() {
+		return this;
+	}
+	async get(name: string): Promise<FluidDataStoreRegistryEntry | undefined> {
+		// this factory is also a registry, which only supports creating itself
+		if (name === this.type) {
+			return this;
+		}
+	}
+
+	get IFluidDataStoreFactory() {
+		return this;
+	}
+	async instantiateDataStore(context, existing) {
+		const runtime: FluidDataStoreRuntime = new FluidDataStoreRuntime(
+			context,
+			sharedObjectRegistry,
+			existing,
+			async () => dataStore,
+		);
+		const dataStore = existing
+			? DataStoreWithSyncCreate.load(context, runtime)
+			: DataStoreWithSyncCreate.create(context, runtime);
+
+		return runtime;
+	}
+}
+
+// a simple container runtime factory with a single datastore aliased as default.
+// the default datastore is also returned as the entrypoint
+const runtimeFactory: IRuntimeFactory = {
+	get IRuntimeFactory() {
+		return this;
+	},
+	instantiateRuntime: async (context, existing) => {
+		return loadContainerRuntime({
+			context,
+			existing,
+			registryEntries: [
+				[
+					DataStoreWithSyncCreateFactory.instance.type,
+					Promise.resolve(DataStoreWithSyncCreateFactory.instance),
+				],
+			],
+			provideEntryPoint: async (rt) => {
+				const maybeRoot = await rt.getAliasedDataStoreEntryPoint("default");
+				if (maybeRoot === undefined) {
+					const ds = await rt.createDataStore(DataStoreWithSyncCreate.type);
+					await ds.trySetAlias("default");
+				}
+				const root = await rt.getAliasedDataStoreEntryPoint("default");
+				assert(root !== undefined, "default must exist");
+				return root.get();
+			},
+		});
+	},
+};
+
+describe("Scenario Test", () => {
+	it("Synchronously create nested data store", async () => {
+		const deltaConnectionServer = LocalDeltaConnectionServer.create();
+
+		const { loader, codeDetails, urlResolver } = createLoader({
+			deltaConnectionServer,
+			runtimeFactory,
+		});
+
+		const container = await loader.createDetachedContainer(codeDetails);
+
+		await container.attach(urlResolver.createCreateNewRequest("test"));
+		const url = await container.getAbsoluteUrl("");
+		assert(url !== undefined, "container must have url");
+		{
+			const entrypoint: FluidObject<DataStoreWithSyncCreate> = await container.getEntryPoint();
+
+			assert(
+				entrypoint.DataStoreWithSyncCreate !== undefined,
+				"container entrypoint must be DataStoreWithSyncCreate",
+			);
+
+			const { attachRuntimeP, dataStore } = entrypoint.DataStoreWithSyncCreate.createAnother();
+
+			dataStore.sharedMap.set("childValue", "childValue");
+
+			// can we make this synchronous
+			await attachRuntimeP;
```

Comment:
Only if the factory retrieval is synchronous. Depends on the scale of the application. Any application that wants to delay load its code or factories, would need a way to work around this. We would need to do that or async load the factories before we create the object. Or use the factories themselves to synchronously create the object.